In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.set_option('display.max_rows', 500)

In [3]:
df = pd.read_csv("bikes.csv", parse_dates=[0], infer_datetime_format=True)

In [4]:
df.head()

datetime  season  holiday  workingday weather  temp  humidity  \
0 2011-01-01 00:00:00  Spring      0.0         0.0   Clear  9.84      81.0   
1 2011-01-01 01:00:00  Spring      0.0         0.0     NaN  9.02      80.0   
2 2011-01-01 02:00:00  Spring      0.0         0.0   Clear  9.02       NaN   
3 2011-01-01 03:00:00  Spring      0.0         0.0   Clear  9.84      75.0   
4 2011-01-01 04:00:00     NaN      0.0         0.0   Clear   NaN      75.0   

   windspeed  casual  registered  rented_bikes_count  
0        NaN       3          13                  16  
1        0.0       8          32                  40  
2        0.0       5          27                  32  
3        0.0       3          10                  13  
4        NaN       0           1                   1

In [5]:
df["weather"].value_counts()

Clear    5793
Mist     2259
Rainy     693
Snowy       1
Name: weather, dtype: int64

In [6]:
df["season"].value_counts()

Winter    2688
Fall      2680
Summer    2670
Spring    2634
Name: season, dtype: int64

In [7]:
df.columns

Index(['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp',
       'humidity', 'windspeed', 'casual', 'registered', 'rented_bikes_count'],
      dtype='object')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   datetime            10886 non-null  datetime64[ns]
 1   season              10672 non-null  object        
 2   holiday             10030 non-null  float64       
 3   workingday          9388 non-null   float64       
 4   weather             8746 non-null   object        
 5   temp                8104 non-null   float64       
 6   humidity            7462 non-null   float64       
 7   windspeed           6820 non-null   float64       
 8   casual              10886 non-null  int64         
 9   registered          10886 non-null  int64         
 10  rented_bikes_count  10886 non-null  int64         
dtypes: datetime64[ns](1), float64(5), int64(3), object(2)
memory usage: 935.6+ KB


In [9]:
df.isna().sum()

datetime                 0
season                 214
holiday                856
workingday            1498
weather               2140
temp                  2782
humidity              3424
windspeed             4066
casual                   0
registered               0
rented_bikes_count       0
dtype: int64

- From info we can see that column datetime is string so we should convert it into date time and make 2 columns seperated one for date and one for time
- we can see also that shape of column is 10886 rows and 11 columns
- workingday column has 1498 null values
- windspeed column has 4066 null values
- temp column has 2782 null values
- humidity column has 3424 null values
- weather column has 2140 null values
- holiday column has 856 null values
- season column has 214 null values

TASKS
----------

1. Make Profit Feature.
    - Each registered user rent bike for 5 USD/hour.
    - Casual user rent bike for 20 USD/hour.
    - Taxes 0.14%.
    - Maintainance 1500/year.
2. Distribution rental_bikes_count & profit.
3. Profit for each day.
4. Which season that has more registration & profit.
5. Which weather condition that has more registration & profit.
6. Correlation between profit & bikes count with all features.
7. Rentals during rush hours (7-9 am & 3-5 pm).
7. Rentals during workday & holiday.
8. Registered or casual is more.
9. What is the average of bikes would be registered during the week.
10. Schools (9-12) & (2-6) profit.
11. Profit in 2011 & Saturdays & season fall

If we used to_datetime we will remove the time part from date time and we will need it

df['datetime']= pd.to_datetime(df['datetime'])



**My approach:-**
- So we restarted the kernel and needed to split the datetime and extract each column alone
- so in the column of time i will turn it into list and take only the first element and this will be number of hours then we will create
- then i will create a profit column one for casual and one for registered

In [10]:
## first we will make 2 new columns for date and time
temp = pd.DatetimeIndex(df['datetime'])
df['date'] = temp.date
df['time'] = temp.time
del df['datetime']

In [11]:
# next we will convert date into datetime type
df["date"] = pd.to_datetime(df["date"])

In [12]:
df.head(1)

season  holiday  workingday weather  temp  humidity  windspeed  casual  \
0  Spring      0.0         0.0   Clear  9.84      81.0        NaN       3   

   registered  rented_bikes_count       date      time  
0          13                  16 2011-01-01  00:00:00

In [13]:
## next we will split time string by ":" and choose first element from list
# df["category"] = df["category"].str.replace('\d+', '')
df["hours"] = df["time"].astype(str).str.replace(":","")

In [14]:
df.head(3)

season  holiday  workingday weather  temp  humidity  windspeed  casual  \
0  Spring      0.0         0.0   Clear  9.84      81.0        NaN       3   
1  Spring      0.0         0.0     NaN  9.02      80.0        0.0       8   
2  Spring      0.0         0.0   Clear  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date      time   hours  
0          13                  16 2011-01-01  00:00:00  000000  
1          32                  40 2011-01-01  01:00:00  010000  
2          27                  32 2011-01-01  02:00:00  020000

In [15]:
df["hours"]= df["hours"].astype(int)

In [16]:
df.head(3)

season  holiday  workingday weather  temp  humidity  windspeed  casual  \
0  Spring      0.0         0.0   Clear  9.84      81.0        NaN       3   
1  Spring      0.0         0.0     NaN  9.02      80.0        0.0       8   
2  Spring      0.0         0.0   Clear  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date      time  hours  
0          13                  16 2011-01-01  00:00:00      0  
1          32                  40 2011-01-01  01:00:00  10000  
2          27                  32 2011-01-01  02:00:00  20000

In [17]:
df["hours"] = df["hours"]/10000

In [18]:
df.head(3)

season  holiday  workingday weather  temp  humidity  windspeed  casual  \
0  Spring      0.0         0.0   Clear  9.84      81.0        NaN       3   
1  Spring      0.0         0.0     NaN  9.02      80.0        0.0       8   
2  Spring      0.0         0.0   Clear  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date      time  hours  
0          13                  16 2011-01-01  00:00:00    0.0  
1          32                  40 2011-01-01  01:00:00    1.0  
2          27                  32 2011-01-01  02:00:00    2.0

In [19]:
## then we will remove time column
df.drop(columns=['time'], inplace = True)

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   season              10672 non-null  object        
 1   holiday             10030 non-null  float64       
 2   workingday          9388 non-null   float64       
 3   weather             8746 non-null   object        
 4   temp                8104 non-null   float64       
 5   humidity            7462 non-null   float64       
 6   windspeed           6820 non-null   float64       
 7   casual              10886 non-null  int64         
 8   registered          10886 non-null  int64         
 9   rented_bikes_count  10886 non-null  int64         
 10  date                10886 non-null  datetime64[ns]
 11  hours               10886 non-null  float64       
dtypes: datetime64[ns](1), float64(6), int64(3), object(2)
memory usage: 1020.7+ KB


In [21]:
## now we will create a column for taxes for each hour
df["taxesperhours"] = (df["hours"] * 0.0014)

In [22]:
df.head(3)

season  holiday  workingday weather  temp  humidity  windspeed  casual  \
0  Spring      0.0         0.0   Clear  9.84      81.0        NaN       3   
1  Spring      0.0         0.0     NaN  9.02      80.0        0.0       8   
2  Spring      0.0         0.0   Clear  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date  hours  taxesperhours  
0          13                  16 2011-01-01    0.0         0.0000  
1          32                  40 2011-01-01    1.0         0.0014  
2          27                  32 2011-01-01    2.0         0.0028

### Task 1:-

- Make Profit Feature.
    - Each registered user rent bike for 5 USD/hour.
    - Casual user rent bike for 20 USD/hour.
    - Taxes 0.14%.
    - Maintainance 1500/year.

- So that means for each registerd user:-
    - (hour * 5) + taxes + 1500
- And for each casual user:-
    - (hour * 20) + taxes 

In [23]:
df["reg_prof"] = (df["hours"]*5) + df["taxesperhours"] + 1500

In [24]:
df["casual_prof"] = (df["hours"] * 20) + df["taxesperhours"]

In [25]:
df.head(5)

season  holiday  workingday weather  temp  humidity  windspeed  casual  \
0  Spring      0.0         0.0   Clear  9.84      81.0        NaN       3   
1  Spring      0.0         0.0     NaN  9.02      80.0        0.0       8   
2  Spring      0.0         0.0   Clear  9.02       NaN        0.0       5   
3  Spring      0.0         0.0   Clear  9.84      75.0        0.0       3   
4     NaN      0.0         0.0   Clear   NaN      75.0        NaN       0   

   registered  rented_bikes_count       date  hours  taxesperhours   reg_prof  \
0          13                  16 2011-01-01    0.0         0.0000  1500.0000   
1          32                  40 2011-01-01    1.0         0.0014  1505.0014   
2          27                  32 2011-01-01    2.0         0.0028  1510.0028   
3          10                  13 2011-01-01    3.0         0.0042  1515.0042   
4           1                   1 2011-01-01    4.0         0.0056  1520.0056   

   casual_prof  
0       0.0000  
1      20.0014  
2      40.0028  
3      60.0042  
4      80.0056

In [26]:
df["weather"].value_counts()

Clear    5793
Mist     2259
Rainy     693
Snowy       1
Name: weather, dtype: int64

In [27]:
di_1 = {"Clear":1, "Mist":2, "Rainy":3, "Snowy":4}
df.replace({"weather": di_1}, inplace = True)

In [28]:
df["workingday"].value_counts()

1.0    6367
0.0    3021
Name: workingday, dtype: int64

In [29]:
df["holiday"].value_counts()

0.0    9738
1.0     292
Name: holiday, dtype: int64

In [30]:
df["season"].value_counts()

Winter    2688
Fall      2680
Summer    2670
Spring    2634
Name: season, dtype: int64

In [31]:
di = {"Winter":1, "Fall":2, "Summer":3, "Spring":4}
df.replace({"season": di}, inplace = True)

In [32]:
df["total_prof"] = df["reg_prof"] + df["casual_prof"]

In [33]:
df.head(3)

season  holiday  workingday  weather  temp  humidity  windspeed  casual  \
0     4.0      0.0         0.0      1.0  9.84      81.0        NaN       3   
1     4.0      0.0         0.0      NaN  9.02      80.0        0.0       8   
2     4.0      0.0         0.0      1.0  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date  hours  taxesperhours   reg_prof  \
0          13                  16 2011-01-01    0.0         0.0000  1500.0000   
1          32                  40 2011-01-01    1.0         0.0014  1505.0014   
2          27                  32 2011-01-01    2.0         0.0028  1510.0028   

   casual_prof  total_prof  
0       0.0000   1500.0000  
1      20.0014   1525.0028  
2      40.0028   1550.0056

#### Example:-

import pandas as pd

df = pd.DataFrame({'ID':['1', '2', '3'], 'col_1': [0, 2, 3], 'col_2':[1, 4, 5]})

mylist = ['a', 'b', 'c', 'd', 'e', 'f']

def get_sublist(sta,end):

    return mylist[sta:end+1]

df['col_3'] = df.apply(lambda x: get_sublist(x.col_1, x.col_2), axis=1)

### Task 2:-
- Distribution rental_bikes_count & profit.

In [34]:
df_new = df[["rented_bikes_count", "total_prof"]]
df_new

rented_bikes_count  total_prof
0                      16   1500.0000
1                      40   1525.0028
2                      32   1550.0056
3                      13   1575.0084
4                       1   1600.0112
...                   ...         ...
10881                 336   1975.0532
10882                 241   2000.0560
10883                 168   2025.0588
10884                 129   2050.0616
10885                  88   2075.0644

[10886 rows x 2 columns]

In [35]:
df_new.describe()

rented_bikes_count    total_prof
count        10886.000000  10886.000000
mean           191.574132   1788.572644
std            181.144454    172.915314
min              1.000000   1500.000000
25%             42.000000   1650.016800
50%            145.000000   1800.033600
75%            284.000000   1950.050400
max            977.000000   2075.064400

So now we have two dataframes one for regular dataframe that contains all columns **df** And the other only contains **rented_bikes_count** and **total_prof** , **new_df**

### Task 3:-
- Profit for each day

our first approach we will only create a column that divide df["total_prof"]/365

but we noticed that there's several profits for one day so we will maybe need to groupby date and sum our first approach

In [36]:
df["prof_per_day"] = df["total_prof"] / 365

In [37]:
df.head(3)

season  holiday  workingday  weather  temp  humidity  windspeed  casual  \
0     4.0      0.0         0.0      1.0  9.84      81.0        NaN       3   
1     4.0      0.0         0.0      NaN  9.02      80.0        0.0       8   
2     4.0      0.0         0.0      1.0  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date  hours  taxesperhours   reg_prof  \
0          13                  16 2011-01-01    0.0         0.0000  1500.0000   
1          32                  40 2011-01-01    1.0         0.0014  1505.0014   
2          27                  32 2011-01-01    2.0         0.0028  1510.0028   

   casual_prof  total_prof  prof_per_day  
0       0.0000   1500.0000      4.109589  
1      20.0014   1525.0028      4.178090  
2      40.0028   1550.0056      4.246591

Maybe we can use this lines of code
- index = pd.MultiIndex.from_arrays(arrays, names=('Animal', 'Type'))
- df.groupby(level=0).sum()

In [38]:
# for profit per date
prperdate_df = df[["date","reg_prof","casual_prof","total_prof","prof_per_day"]]
prperdate_df.groupby('date').sum()

reg_prof  casual_prof  total_prof  prof_per_day
date                                                         
2011-01-01  37380.3864    5520.3864  42900.7728    117.536364
2011-01-02  35855.3794    5420.3794  41275.7588    113.084271
2011-01-03  34355.3794    5420.3794  39775.7588    108.974682
2011-01-04  35865.3822    5460.3822  41325.7644    113.221272
2011-01-05  35865.3822    5460.3822  41325.7644    113.221272
2011-01-06  35865.3822    5460.3822  41325.7644    113.221272
2011-01-07  35865.3822    5460.3822  41325.7644    113.221272
2011-01-08  37380.3864    5520.3864  42900.7728    117.536364
2011-01-09  37380.3864    5520.3864  42900.7728    117.536364
2011-01-10  37380.3864    5520.3864  42900.7728    117.536364
2011-01-11  34345.3766    5380.3766  39725.7532    108.837680
2011-01-12  34345.3766    5380.3766  39725.7532    108.837680
2011-01-13  37380.3864    5520.3864  42900.7728    117.536364
2011-01-14  35860.3808    5440.3808  41300.7616    113.152772
2011-01-15  37380.3864    5520.3864  42900.7728    117.536364
2011-01-16  37380.3864    5520.3864  42900.7728    117.536364
2011-01-17  37380.3864    5520.3864  42900.7728    117.536364
2011-01-18  19050.2940    4200.2940  23250.5880     63.700241
2011-01-19  35865.3822    5460.3822  41325.7644    113.221272
2011-02-01  35860.3808    5440.3808  41300.7616    113.152772
2011-02-02  37380.3864    5520.3864  42900.7728    117.536364
2011-02-03  35860.3808    5440.3808  41300.7616    113.152772
2011-02-04  35860.3808    5440.3808  41300.7616    113.152772
2011-02-05  37380.3864    5520.3864  42900.7728    117.536364
2011-02-06  37380.3864    5520.3864  42900.7728    117.536364
2011-02-07  37380.3864    5520.3864  42900.7728    117.536364
2011-02-08  37380.3864    5520.3864  42900.7728    117.536364
2011-02-09  35860.3808    5440.3808  41300.7616    113.152772
2011-02-10  35865.3822    5460.3822  41325.7644    113.221272
2011-02-11  34345.3766    5380.3766  39725.7532    108.837680
2011-02-12  37380.3864    5520.3864  42900.7728    117.536364
2011-02-13  35855.3794    5420.3794  41275.7588    113.084271
2011-02-14  37380.3864    5520.3864  42900.7728    117.536364
2011-02-15  35865.3822    5460.3822  41325.7644    113.221272
2011-02-16  35870.3836    5480.3836  41350.7672    113.289773
2011-02-17  37380.3864    5520.3864  42900.7728    117.536364
2011-02-18  37380.3864    5520.3864  42900.7728    117.536364
2011-02-19  37380.3864    5520.3864  42900.7728    117.536364
2011-03-01  37380.3864    5520.3864  42900.7728    117.536364
2011-03-02  37380.3864    5520.3864  42900.7728    117.536364
2011-03-03  37380.3864    5520.3864  42900.7728    117.536364
2011-03-04  37380.3864    5520.3864  42900.7728    117.536364
2011-03-05  37380.3864    5520.3864  42900.7728    117.536364
2011-03-06  35855.3794    5420.3794  41275.7588    113.084271
2011-03-07  35870.3836    5480.3836  41350.7672    113.289773
2011-03-08  37380.3864    5520.3864  42900.7728    117.536364
2011-03-09  37380.3864    5520.3864  42900.7728    117.536364
2011-03-10  34345.3766    5380.3766  39725.7532    108.837680
2011-03-11  35860.3808    5440.3808  41300.7616    113.152772
2011-03-12  37380.3864    5520.3864  42900.7728    117.536364
2011-03-13  35870.3836    5480.3836  41350.7672    113.289773
2011-03-14  35860.3808    5440.3808  41300.7616    113.152772
2011-03-15  35865.3822    5460.3822  41325.7644    113.221272
2011-03-16  35870.3836    5480.3836  41350.7672    113.289773
2011-03-17  37380.3864    5520.3864  42900.7728    117.536364
2011-03-18  35860.3808    5440.3808  41300.7616    113.152772
2011-03-19  37380.3864    5520.3864  42900.7728    117.536364
2011-04-01  37380.3864    5520.3864  42900.7728    117.536364
2011-04-02  37380.3864    5520.3864  42900.7728    117.536364
2011-04-03  37380.3864    5520.3864  42900.7728    117.536364
2011-04-04  37380.3864    5520.3864  42900.7728    117.536364
2011-04-05  37380.3864    5520.3864  42900.7728    117.536364
2011-04-06  37380.3864    5520.3864  42900.772

### Task 4:-
- Which season that has more registration & profit.
    - Winter = 1
    - Fall = 2
    - Summer = 3
    - Spring = 4
    
our approach we can group by season in a new dataframe and get statistics from them

In [39]:
season_profit_df = df[["season","registered","reg_prof","casual_prof","total_prof","prof_per_day"]]
season_profit_df.groupby('season').sum()

registered      reg_prof  casual_prof    total_prof  prof_per_day
season                                                                   
1.0         457276  4.186218e+06  616743.1690  4.802961e+06  13158.798186
2.0         488252  4.174223e+06  616763.1704  4.790986e+06  13125.989975
3.0         446991  4.158238e+06  612822.8946  4.771061e+06  13071.399422
4.0         265322  4.104403e+06  613482.9408  4.717886e+06  12925.714744

As we can see the total prof is the highest total prof is at **1st** season which is **Winter** While the most registered season is at **3rd** season which is **Summer**

### Task 5:-
- Which weather condition that has more registration & profit.

We will do the same as task 4 but this time is for weather:-

- Clear = 1
- Mist = 2
- Rainy = 3
- Snowy = 4

In [40]:
weather_profit_df = df[["weather","registered","total_prof","prof_per_day"]]
weather_profit_df.groupby('weather').sum()

registered    total_prof  prof_per_day
weather                                        
1.0          951449  1.038139e+07  28442.162933
2.0          336229  4.002694e+06  10966.284334
3.0           70204  1.244623e+06   3409.925948
4.0             158  1.950050e+03      5.342604

As we can see there's a max total profit and total registeration happen at **1st** weather which is **Clear**

### Task 6:-
- Correlation between profit & bikes count with all features.

In [41]:
df[["total_prof","rented_bikes_count"]].corr()

total_prof  rented_bikes_count
total_prof            1.000000            0.400601
rented_bikes_count    0.400601            1.000000

In [42]:
## Another way to tell how they correlate to each other
column_1 = df["total_prof"]
column_2 = df["rented_bikes_count"]
correlation = column_1.corr(column_2)
print(correlation)

0.40060119414684703


In [43]:
df.corr()

season   holiday  workingday   weather      temp  \
season              1.000000 -0.028007    0.008605 -0.011242 -0.270544   
holiday            -0.028007  1.000000   -0.249189 -0.011779 -0.001047   
workingday          0.008605 -0.249189    1.000000  0.035717  0.018954   
weather            -0.011242 -0.011779    0.035717  1.000000 -0.049568   
temp               -0.270544 -0.001047    0.018954 -0.049568  1.000000   
humidity           -0.186729  0.016190   -0.019167  0.380920 -0.068240   
windspeed           0.162669  0.010463    0.020668 -0.002799 -0.025425   
casual             -0.097876  0.042519   -0.318348 -0.132852  0.466035   
registered         -0.163983 -0.021017    0.118612 -0.106523  0.315430   
rented_bikes_count -0.163734 -0.005827    0.010940 -0.125342  0.391867   
hours               0.007862  0.001522    0.004650 -0.022797  0.147415   
taxesperhours       0.007862  0.001522    0.004650 -0.022797  0.147415   
reg_prof            0.007862  0.001522    0.004650 -0.022797  0.147415   
casual_prof         0.007862  0.001522    0.004650 -0.022797  0.147415   
total_prof          0.007862  0.001522    0.004650 -0.022797  0.147415   
prof_per_day        0.007862  0.001522    0.004650 -0.022797  0.147415   

                    humidity  windspeed    casual  registered  \
season             -0.186729   0.162669 -0.097876   -0.163983   
holiday             0.016190   0.010463  0.042519   -0.021017   
workingday         -0.019167   0.020668 -0.318348    0.118612   
weather             0.380920  -0.002799 -0.132852   -0.106523   
temp               -0.068240  -0.025425  0.466035    0.315430   
humidity            1.000000  -0.312628 -0.344939   -0.266432   
windspeed          -0.312628   1.000000  0.085055    0.089105   
casual             -0.344939   0.085055  1.000000    0.497250   
registered         -0.266432   0.089105  0.497250    1.000000   
rented_bikes_count -0.317282   0.097747  0.690414    0.970948   
hours              -0.287224   0.145908  0.302045    0.380540   
taxesperhours      -0.287224   0.145908  0.302045    0.380540   
reg_prof           -0.287224   0.145908  0.302045    0.380540   
casual_prof        -0.287224   0.145908  0.302045    0.380540   
total_prof         -0.287224   0.145908  0.302045    0.380540   
prof_per_day       -0.287224   0.145908  0.302045    0.380540   

                    rented_bikes_count     hours  taxesperhours  reg_prof  \
season                       -0.163734  0.007862       0.007862  0.007862   
holiday                      -0.005827  0.001522       0.001522  0.001522   
workingday                    0.010940  0.004650       0.004650  0.004650   
weather                      -0.125342 -0.022797      -0.022797 -0.022797   
temp                          0.391867  0.147415       0.147415  0.147415   
humidity                     -0.317282 -0.287224      -0.287224 -0.287224   
windspeed                     0.097747  0.145908       0.145908  0.145908   
casual                        0.690414  0.302045       0.302045  0.302045   
registered                    0.970948  0.380540       0.380540  0.380540   
rented_bikes_count            1.000000  0.400601       0.400601  0.400601   
hours                         0.400601  1.000000       1.000000  1.000000   
taxesperhours                 0.400601  1.000000       1.000000  1.000000   
reg_prof                      0.400601  1.000000       1.000000  1.000000   
casual_prof                   0.400601  1.000000       1.000000  1.000000   
total_prof                    0.400601  1.000000       1.000000  1.000000   
prof_per_day                  0.400601  1.000000       1.000000  1.000000   

                    casual_prof  total_prof  prof_per_day  
season                 0.007862    0.007862      0.007862  
holiday                0.001522    0.001522      0.001522  
workingday             0.004650    0.004650      0.004650  
weather               -0.022797   -0.022797     -0.022797  
temp                   0.147415    0.147415    

In [44]:
df.columns

Index(['season', 'holiday', 'workingday', 'weather', 'temp', 'humidity',
       'windspeed', 'casual', 'registered', 'rented_bikes_count', 'date',
       'hours', 'taxesperhours', 'reg_prof', 'casual_prof', 'total_prof',
       'prof_per_day'],
      dtype='object')

In [45]:
table = pd.pivot_table(df, index=['rented_bikes_count', 'total_prof']).corr()

In [46]:
table

casual  casual_prof   holiday     hours  humidity  \
casual         1.000000     0.178522  0.050336  0.178522 -0.355014   
casual_prof    0.178522     1.000000 -0.003965  1.000000 -0.252773   
holiday        0.050336    -0.003965  1.000000 -0.003965  0.024561   
hours          0.178522     1.000000 -0.003965  1.000000 -0.252773   
humidity      -0.355014    -0.252773  0.024561 -0.252773  1.000000   
prof_per_day   0.178522     1.000000 -0.003965  1.000000 -0.252773   
reg_prof       0.178522     1.000000 -0.003965  1.000000 -0.252773   
registered     0.408139     0.188372 -0.032020  0.188372 -0.212639   
season        -0.142318     0.034732 -0.031619  0.034732 -0.155534   
taxesperhours  0.178522     1.000000 -0.003965  1.000000 -0.252773   
temp           0.505393     0.116944  0.028287  0.116944 -0.143677   
weather       -0.184587    -0.029161  0.004323 -0.029161  0.382473   
windspeed      0.045060     0.118807  0.004172  0.118807 -0.267109   
workingday    -0.379285     0.085961 -0.222441  0.085961 -0.029233   

               prof_per_day  reg_prof  registered    season  taxesperhours  \
casual             0.178522  0.178522    0.408139 -0.142318       0.178522   
casual_prof        1.000000  1.000000    0.188372  0.034732       1.000000   
holiday           -0.003965 -0.003965   -0.032020 -0.031619      -0.003965   
hours              1.000000  1.000000    0.188372  0.034732       1.000000   
humidity          -0.252773 -0.252773   -0.212639 -0.155534      -0.252773   
prof_per_day       1.000000  1.000000    0.188372  0.034732       1.000000   
reg_prof           1.000000  1.000000    0.188372  0.034732       1.000000   
registered         0.188372  0.188372    1.000000 -0.234420       0.188372   
season             0.034732  0.034732   -0.234420  1.000000       0.034732   
taxesperhours      1.000000  1.000000    0.188372  0.034732       1.000000   
temp               0.116944  0.116944    0.355993 -0.313956       0.116944   
weather           -0.029161 -0.029161   -0.165042  0.012421      -0.029161   
windspeed          0.118807  0.118807    0.036688  0.171113       0.118807   
workingday         0.085961  0.085961    0.204290  0.022209       0.085961   

                   temp   weather  windspeed  workingday  
casual         0.505393 -0.184587   0.045060   -0.379285  
casual_prof    0.116944 -0.029161   0.118807    0.085961  
holiday        0.028287  0.004323   0.004172   -0.222441  
hours          0.116944 -0.029161   0.118807    0.085961  
humidity      -0.143677  0.382473  -0.267109   -0.029233  
prof_per_day   0.116944 -0.029161   0.118807    0.085961  
reg_prof       0.116944 -0.029161   0.118807    0.085961  
registered     0.355993 -0.165042   0.036688    0.204290  
season        -0.313956  0.012421   0.171113    0.022209  
taxesperhours  0.116944 -0.029161   0.118807    0.085961  
temp           1.000000 -0.105438  -0.026427   -0.008733  
weather       -0.105438  1.000000  -0.006056    0.041020  
windspeed     -0.026427 -0.006056   1.000000    0.033855  
workingday    -0.008733  0.041020   0.033855    1.000000

### Task 7:-
- Rentals during rush hours (7-9 am & 3-5 pm).

our approach:-
- sum of rented bike counts from 7 - 9 am ----> which means from **hours = 7** till **hourse = 9** in our modified dataframe
- sum of rented bike counts from 3 - 5 pm ----> which means from **hours = 15** till **hours = 17** in our modified dataframe

**What we will use**
- rented_bikes_count
- hours between 7 to 9 and hours between 15 to 17

In [47]:
new = df.groupby('hours')[['rented_bikes_count']].sum()

In [48]:
new

rented_bikes_count
hours                    
0.0                 25088
1.0                 15372
2.0                 10259
3.0                  5091
4.0                  2832
5.0                  8935
6.0                 34698
7.0                 96968
8.0                165060
9.0                100910
10.0                79667
11.0                95857
12.0               116968
13.0               117551
14.0               111010
15.0               115960
16.0               144266
17.0               213757
18.0               196472
19.0               143767
20.0               104204
21.0                79057
22.0                60911
23.0                40816

- 7.0 	96968
- 8.0 	165060
- 9.0 	100910

And
- 15.0 	115960
- 16.0 	144266
- 17.0 	213757

In [49]:
## simple way in our case:-
dumb_sum =  96968 + 165060 + 100910 + 115960 + 144266 + 213757
print(dumb_sum)

836921


let's do an advanced way of sum values that get values from  hours = 7.0 to hours = 9.0 and from hours = 15.0 to hours = 17.0

In [50]:
# sum_1 = df.iloc[(df['hours'] == 7.0) & (df['hours'] == 9.0), 'rented_bikes_count'].sum()
# the problem with this way is that we need to make hours as index to select and sum
## so we will do this way
hours_rent_df = df[["hours", "rented_bikes_count"]]
hours_rent_df.groupby('hours').sum()

rented_bikes_count
hours                    
0.0                 25088
1.0                 15372
2.0                 10259
3.0                  5091
4.0                  2832
5.0                  8935
6.0                 34698
7.0                 96968
8.0                165060
9.0                100910
10.0                79667
11.0                95857
12.0               116968
13.0               117551
14.0               111010
15.0               115960
16.0               144266
17.0               213757
18.0               196472
19.0               143767
20.0               104204
21.0                79057
22.0                60911
23.0                40816

### Task 8:-
- Rentals during workday & holiday.

our approach:-
- sum of rented bike count when workday = 1
- sum of rented bike count when holiday = 1

In [51]:
df.loc[df['workingday'] == 1, 'rented_bikes_count'].sum()

1226368

In [52]:
df.loc[df['holiday'] == 1, 'rented_bikes_count'].sum()

54128

### Task 9:-
- Registered or casual is more.

our approach:-
- value_counts of each column and get sum of values

In [53]:
df["casual"].sum()

392135

In [54]:
df["registered"].sum()

1693341

**Registered is more than casual**

### Task 10:-
- What is the average of bikes would be registered during the week.

our approach:-
- get dates mean from first date until 7th date
    - it maybe false and we need to get mean for every 7 days because it may differ from week to week then we get the mean

In [55]:
df.iloc[0:7,8].mean()

12.0

In [56]:
df.iloc[8:15,8].mean()

34.285714285714285

In [57]:
df.iloc[16:23,8].mean()

33.42857142857143

**12 bikes for first week and then it differes between 34 and 33 per week**

### Task 11:-
- Schools (9-12) & (2-6) profit.
    - Maybe it means months

In [59]:
df['month'] = df['date'].dt.month

In [60]:
df.head(3)

season  holiday  workingday  weather  temp  humidity  windspeed  casual  \
0     4.0      0.0         0.0      1.0  9.84      81.0        NaN       3   
1     4.0      0.0         0.0      NaN  9.02      80.0        0.0       8   
2     4.0      0.0         0.0      1.0  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date  hours  taxesperhours   reg_prof  \
0          13                  16 2011-01-01    0.0         0.0000  1500.0000   
1          32                  40 2011-01-01    1.0         0.0014  1505.0014   
2          27                  32 2011-01-01    2.0         0.0028  1510.0028   

   casual_prof  total_prof  prof_per_day  month  
0       0.0000   1500.0000      4.109589      1  
1      20.0014   1525.0028      4.178090      1  
2      40.0028   1550.0056      4.246591      1

In [71]:
## this is for all months we calculate total profits for each
df.groupby(df["month"])["total_prof"].sum()

month
1     1.585279e+06
2     1.612754e+06
3     1.612854e+06
4     1.625479e+06
5     1.630229e+06
6     1.630229e+06
7     1.630229e+06
8     1.630229e+06
9     1.625579e+06
10    1.628654e+06
11    1.628654e+06
12    1.630229e+06
Name: total_prof, dtype: float64

In [77]:
sumdumb_from_9_to_12 = 1.625579e+06 + 1.628654e+06 + 1.628654e+06 + 1.630229e+06
sumdumb_from_9_to_12

6513116.0

In [78]:
sumdumb_from_2_to_6 =1.612754e+06 + 1.612854e+06 + 1.625479e+06 + 1.630229e+06 + 1.630229e+06
sumdumb_from_2_to_6

8111545.0

**Please add comment here and let me know how to sum on conditions**

In [80]:
# df.groupby('date').apply(lambda x: x.resample('7D', on='date').sum())

### Task 12:-
- Profit in 2011 & Saturdays & season fall

our approach:-
- select date where year = 2011
- select from date where we should define a list of weekday names
    - like this
        - days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
        - days[date.weekday()]
    - or we should use this:-
        - df['Date'].dt.day_name()
- select season = fall or in our dataframe == 2

In [81]:
df['year'] = df['date'].dt.year

In [82]:
df.head(1)

season  holiday  workingday  weather  temp  humidity  windspeed  casual  \
0     4.0      0.0         0.0      1.0  9.84      81.0        NaN       3   

   registered  rented_bikes_count       date  hours  taxesperhours  reg_prof  \
0          13                  16 2011-01-01    0.0            0.0    1500.0   

   casual_prof  total_prof  prof_per_day  month  year  
0          0.0      1500.0      4.109589      1  2011

In [83]:
df["day_name"] = df['date'].dt.day_name()

In [84]:
df.head(3)

season  holiday  workingday  weather  temp  humidity  windspeed  casual  \
0     4.0      0.0         0.0      1.0  9.84      81.0        NaN       3   
1     4.0      0.0         0.0      NaN  9.02      80.0        0.0       8   
2     4.0      0.0         0.0      1.0  9.02       NaN        0.0       5   

   registered  rented_bikes_count       date  hours  taxesperhours   reg_prof  \
0          13                  16 2011-01-01    0.0         0.0000  1500.0000   
1          32                  40 2011-01-01    1.0         0.0014  1505.0014   
2          27                  32 2011-01-01    2.0         0.0028  1510.0028   

   casual_prof  total_prof  prof_per_day  month  year  day_name  
0       0.0000   1500.0000      4.109589      1  2011  Saturday  
1      20.0014   1525.0028      4.178090      1  2011  Saturday  
2      40.0028   1550.0056      4.246591      1  2011  Saturday

In [89]:
profit_on_conditions = df.loc[(df['year']==2011) & (df['day_name'] == "Saturday") & (df['season'] == 2.0),['total_prof']].sum()

In [90]:
print(profit_on_conditions)

total_prof    334681.0676
dtype: float64


Total profits = **334681.0676**